### PR TITLE
NFS over REST

### PR DESCRIPTION
**Pull request by seanmcc-msft**
TODO
- Fix existing CI
- Remember to update Changelogs breaking changes section

### CODE REVIEW

File: sdk/storage/Azure.Storage.Files.Shares/src/Models/NfsFileMode.cs [hunk 1]
```diff
@@ -0,0 +1,43 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+namespace Azure.Storage.Files.Shares.Models
+{
+    /// <summary>
+    /// Only applicable to NFS files and directories.
+    /// The mode permissions of the file or directory.
+    /// </summary>
+    public class NfsFileMode
```
Comment (seanmcc-msft) [inline sdk/storage/Azure.Storage.Files.Shares/src/Models/NfsFileMode.cs:10]: Consider adding ToSymbolicPermissions(), ToOctalPermissions(), ParseOctalPermsisions() and ParseSymbolicPermissions(), similar to Data Lake PathPermissions()

File: sdk/storage/Azure.Storage.Files.Shares/src/Models/NfsFileMode.cs [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+            return stringBuilder.ToString();
+        }
+
+        /// <summary>
+        /// Returns a <see cref="NfsFileMode"/> from the octal string representation.
+        /// </summary>
+        /// <param name="modeString">
+        /// A 4-digit octal string representation of a File Mode.
+        /// </param>
+        public static NfsFileMode ParseOctalFileMode(string modeString)
```
Comment (seanmcc-msft) [inline sdk/storage/Azure.Storage.Files.Shares/src/Models/NfsFileMode.cs:85]: TODO write tests to round-trip file mode with ToOctalFileMode()

File: sdk/storage/Azure.Storage.Files.Shares/src/Models/NfsFileMode.cs [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+        {
```
Comment (seanmcc-msft) [inline sdk/storage/Azure.Storage.Files.Shares/src/Models/NfsFileMode.cs:86]: TODO add additional input validation to this method.

File: sdk/storage/Azure.Storage.Files.Shares/src/Models/NfsFileMode.cs [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+            }
+            return stringBuilder.ToString();
+        }
+        /// <summary>
+        /// Returns a <see cref="NfsFileMode"/> from the symbolic string representation.
+        /// </summary>
+        /// <param name="modeString">
+        /// A 9-character symbolic string representation of a File Mode.
+        /// </param>
+        public static NfsFileMode ParseSymbolicFileMode(string modeString)
```
Comment (seanmcc-msft) [inline sdk/storage/Azure.Storage.Files.Shares/src/Models/NfsFileMode.cs:179]: TODO write tests to round-trip file mode with ToSymbolicFileMode()

File: sdk/storage/Azure.Storage.Files.Shares/src/Models/NfsFileMode.cs [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+        {
```
Comment (seanmcc-msft) [inline sdk/storage/Azure.Storage.Files.Shares/src/Models/NfsFileMode.cs:180]: TODO add additional input validation to this method.

File: sdk/storage/Azure.Storage.Files.Shares/src/ShareFileClient.cs [hunk 6]
```diff
@@ -965,21 +975,20 @@ private async Task<Response<ShareFileInfo>> CreateInternal(
...
-                            fileCreationTime: smbProps.FileCreatedOn.ToFileDateTimeString() ?? Constants.File.FileTimeNow,
-                            fileLastWriteTime: smbProps.FileLastWrittenOn.ToFileDateTimeString() ?? Constants.File.FileTimeNow,
+                            fileAttributes: smbProps.FileAttributes.ToAttributesString(),
+                            fileCreationTime: smbProps.FileCreatedOn.ToFileDateTimeString(),
+                            fileLastWriteTime: smbProps.FileLastWrittenOn.ToFileDateTimeString(),
                             fileChangeTime: smbProps.FileChangedOn.ToFileDateTimeString(),
+                            owner: nfsProperties?.Owner,
+                            group: nfsProperties?.Group,
+                            fileMode: nfsProperties?.FileMode.ToOctalFileMode(),
+                            nfsFileType: nfsProperties?.FileType,
```
[the remark this answers is not in the document]
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Files.Shares/src/ShareFileClient.cs:991]: Remove this line.

File: sdk/storage/Azure.Storage.Files.Shares/src/ShareFileClient.cs [hunk 7]
```diff
@@ -993,10 +1002,14 @@ private async Task<Response<ShareFileInfo>> CreateInternal(
...
```
Comment (seanmcc-msft) [inline sdk/storage/Azure.Storage.Files.Shares/src/ShareFileClient.cs:1012]: Remove this line.

File: sdk/storage/Azure.Storage.Files.Shares/src/Models/FileNfsProperties.cs [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+    {
+        /// <summary>
+        /// Optional. Version TBD and newer. The mode permissions to be set on the file or directory.
+        /// </summary>
+        public NfsFileMode FileMode { get; set; }
+
+        /// <summary>
+        /// Optional. The owner user identifier (UID) to be set on the file or directory. The default value is 0 (root).
+        /// </summary>
+        public uint? Owner { get; set; }
```
Comment (seanmcc-msft) [inline sdk/storage/Azure.Storage.Files.Shares/src/Models/FileNfsProperties.cs:26]: Change to string to pattern match Data Lake

File: sdk/storage/Azure.Storage.Files.Shares/src/Models/FileNfsProperties.cs [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+
+        /// <summary>
+        /// Optional. The owner group identifier (GID) to be set on the file or directory. The default value is 0 (root group).
+        /// </summary>
+        public uint? Group { get; set; }
```
Comment (seanmcc-msft) [inline sdk/storage/Azure.Storage.Files.Shares/src/Models/FileNfsProperties.cs:31]: Change to string to pattern match Data Lake

File: sdk/storage/Azure.Storage.Files.Shares/src/Models/ShareFileCopyOptions.cs [hunk 10]
```diff
@@ -64,5 +64,22 @@ public class ShareFileCopyOptions
...
+        /// <summary>
+        /// Only applicable to NFS Files.  NFS properties to set on the destination file.
+        /// </summary>
+        public FileNfsProperties NfsProperties { get; set; }
+
+        /// <summary>
+        /// Only applicable to NFS Files.  Specifies if the the mode should be copied from the sourece
+        /// to destination file, or overwritten with a new value for the destination.
+        /// </summary>
+        public ModeCopyMode? ModeCopyMode { get; set; }
```
Comment (seanmcc-msft) [inline sdk/storage/Azure.Storage.Files.Shares/src/Models/ShareFileCopyOptions.cs:77]: TODO depending on what the Files team says, we may need to change this XML comment to say this parameter is mandatory for NFS files.

File: sdk/storage/Azure.Storage.Files.Shares/src/Models/ShareFileCopyOptions.cs [hunk 11]
```diff
@@ -64,5 +64,22 @@ public class ShareFileCopyOptions
...
+
+        /// <summary>
+        /// Only applicable to NFS Files.  Specifies if the the owner should be copied from the sourece
+        /// to destination file, or overwritten with a new value for the destination.
+        /// </summary>
+        public OwnerCopyMode? OwnerCopyMode { get; set; }
```
Comment (seanmcc-msft) [inline sdk/storage/Azure.Storage.Files.Shares/src/Models/ShareFileCopyOptions.cs:83]: TODO depending on what the Files team says, we may need to change this XML comment to say this parameter is mandatory for NFS files.

File: sdk/storage/Azure.Storage.Files.Shares/src/ShareFileClient.cs [hunk 12]
```diff
@@ -1770,6 +1815,9 @@ private async Task<Response<ShareFileCopyInfo>> StartCopyInternal(
                         SetArchiveAttribute = setArchiveAttribute
                     };
 
+                    modeCopyMode ??= ModeCopyMode.Source;
+                    ownerCopyMode ??= OwnerCopyMode.Source;
```
Comment (seanmcc-msft) [inline sdk/storage/Azure.Storage.Files.Shares/src/ShareFileClient.cs:1819]: TODO, if these are required for NFS but not allowed for SMB, we should remove these two lines.

Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Files.Shares/src/ShareFileClient.cs]: Fixed.

File: sdk/storage/Azure.Storage.Files.Shares/src/Models/ShareFileCopyOptions.cs (hunk 11 above)
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Files.Shares/src/Models/ShareFileCopyOptions.cs]: Fixed.

File: sdk/storage/Azure.Storage.Files.Shares/src/Models/ShareFileCopyOptions.cs (hunk 10 above)
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Files.Shares/src/Models/ShareFileCopyOptions.cs]: Fixed.

File: sdk/storage/Azure.Storage.Files.Shares/src/Models/NfsFileMode.cs [hunk 13]
```diff
@@ -0,0 +1,206 @@
...
+using System.Globalization;
+using System.Text;
+
+namespace Azure.Storage.Files.Shares.Models
+{
+    /// <summary>
+    /// Only applicable to NFS files and directories.
+    /// The mode permissions of the file or directory.
+    /// </summary>
+    public class NfsFileMode
```
Comment (seanmcc-msft) [inline sdk/storage/Azure.Storage.Files.Shares/src/Models/NfsFileMode.cs:14]: TODO create comparison methods for this class

**Comment by seanmcc-msft** [comment]
TODO may need to make Create Symbolic Link, Get Symbolic Link, and Create Hard Link internal if the Files team moves these to STG 98

File: sdk/storage/Azure.Storage.Files.Shares/src/Models/FileNfsProperties.cs [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+
+        /// <summary>
+        /// Optional. The owner user identifier (UID) to be set on the file or directory. The default value is 0 (root).
+        /// </summary>
+        public string Owner { get; set; }
+
+        /// <summary>
+        /// Optional. The owner group identifier (GID) to be set on the file or directory. The default value is 0 (root group).
+        /// </summary>
+        public string Group { get; set; }
```
Comment (seanmcc-msft) [inline sdk/storage/Azure.Storage.Files.Shares/src/Models/FileNfsProperties.cs:31]: TODO comment for Arch Board on API view about the type of this property.

File: sdk/storage/Azure.Storage.Files.Shares/src/Models/FileNfsProperties.cs [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+    {
+        /// <summary>
+        /// Optional. Version TBD and newer. The mode permissions to be set on the file or directory.
+        /// </summary>
+        public NfsFileMode FileMode { get; set; }
+
+        /// <summary>
+        /// Optional. The owner user identifier (UID) to be set on the file or directory. The default value is 0 (root).
+        /// </summary>
+        public string Owner { get; set; }
```
Comment (seanmcc-msft) [inline sdk/storage/Azure.Storage.Files.Shares/src/Models/FileNfsProperties.cs:26]: TODO comment for Arch Board on API view about the type of this property.

**Comment by seanmcc-msft** [comment]
TODO make Create Symbolic Link and Get Symbolic Link internal

File: sdk/storage/Azure.Storage.Files.Shares/src/Models/NfsFileMode.cs (hunk 13 above)
Comment (seanmcc-msft) [inline sdk/storage/Azure.Storage.Files.Shares/src/Models/NfsFileMode.cs:14]: TODO consider renaming this `PosixFileMode`.  Bring this up with Arch Board.

File: sdk/storage/Azure.Storage.Files.Shares/src/ShareFileClient.cs [hunk 16]
```diff
@@ -6660,6 +6718,424 @@ private async Task<Response<ShareFileClient>> RenameInternal(
...
+        /// A <see cref="Response{StorageFileInfo}"/> describing the
+        /// state of the file.
+        /// </returns>
+        /// <remarks>
+        /// A <see cref="RequestFailedException"/> will be thrown if
+        /// a failure occurs.
+        /// </remarks>
+        /// https://github.com/Azure/azure-sdk-for-net/issues/46907
+        internal virtual Response<ShareFileInfo> CreateSymbolicLink(
+            string path,
```
Comment (seanmcc-msft) [inline sdk/storage/Azure.Storage.Files.Shares/src/ShareFileClient.cs:6857]: TODO this needs to be `LinkText` to align with the Arch Board and Files Team feedback

File: sdk/storage/Azure.Storage.Files.Shares/src/Models/NfsFileMode.cs (hunk 13 above)
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Files.Shares/src/Models/NfsFileMode.cs]: Done

File: sdk/storage/Azure.Storage.Files.Shares/src/autorest.md [hunk 17]
```diff
@@ -4,7 +4,7 @@ Run `dotnet build /t:GenerateCode` to generate code.
 
 ``` yaml
 input-file:
-    - https://raw.githubusercontent.com/Azure/azure-rest-api-specs/ae95eb6a4701d844bada7d1c4f5ecf4a7444e5b8/specification/storage/data-plane/Microsoft.FileStorage/stable/2025-01-05/file.json
+    - C:\azure-rest-api-specs\specification\storage\data-plane\Microsoft.FileStorage\stable\2025-05-05\file.json
```
[the remark this answers is not in the document]
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Files.Shares/src/autorest.md]: Done

**Review by amnguye — APPROVED** [review]
Had a few comments and a question

File: sdk/storage/Azure.Storage.Files.Shares/tests/FileClientTests.cs [hunk 18]
```diff
@@ -10,13 +10,16 @@
 using System.Threading;
 using System.Threading.Tasks;
 using Azure.Core.TestFramework;
+using Azure.Identity;
 using Azure.Storage.Files.Shares.Models;
 using Azure.Storage.Files.Shares.Specialized;
 using Azure.Storage.Sas;
 using Azure.Storage.Test;
 using Azure.Storage.Test.Shared;
+using Microsoft.Identity.Client;
```
[the remark this answers is not in the document]
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Files.Shares/tests/FileClientTests.cs:19]: Fixed.

File: sdk/storage/Azure.Storage.Files.Shares/src/Models/RolePermissionExtensions.cs [hunk 19]
```diff
@@ -0,0 +1,165 @@
...
+                throw new ArgumentException($"Invalid character in symbolic role permission: {s[1]}");
+            }
+
+            // Execute character
+            if (s[2] == 'x' || s[2] == 's' || s[2] == 't')
+            {
+                rolePermissions |= RolePermissions.Execute;
+                if (s[2] == 's' || s[2] == 't')
+                {
+                    setSticky = true;
```
[the remark this answers is not in the document]
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Files.Shares/src/Models/RolePermissionExtensions.cs:149]: I am handling SitckyBit in `NfsFileMode`

File: sdk/storage/Azure.Storage.Files.Shares/src/ShareDirectoryClient.cs [hunk 20]
```diff
@@ -754,38 +762,40 @@ internal async Task<Response<ShareDirectoryInfo>> CreateInternal(
...
                     ResponseWithHeaders<DirectoryCreateHeaders> response;
 
                     if (async)
                     {
                         response = await DirectoryRestClient.CreateAsync(
-                            fileAttributes: smbProps.FileAttributes?.ToAttributesString() ?? Constants.File.FileAttributesNone,
-                            fileCreationTime: smbProps.FileCreatedOn.ToFileDateTimeString() ?? Constants.File.FileTimeNow,
-                            fileLastWriteTime: smbProps.FileLastWrittenOn.ToFileDateTimeString() ?? Constants.File.FileTimeNow,
+                            fileAttributes: smbProps?.FileAttributes.ToAttributesString(),
+                            fileCreationTime: smbProps?.FileCreatedOn.ToFileDateTimeString(),
```
[the remark this answers is not in the document]
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Files.Shares/src/ShareDirectoryClient.cs:772]: Adding the `?` after `FileAttributes`, `FileCreatedOn`, and `FileLastWrittenOn` results in a compile error.

File: sdk/storage/Azure.Storage.Files.Shares/src/ShareFileClient.cs [hunk 21]
```diff
@@ -129,7 +130,7 @@ public virtual string Path
         /// </summary>
         public virtual bool CanGenerateSasUri => ClientConfiguration.SharedKeyCredential != null;
 
-        //const string fileType = "file";
+        //const string FileType = "file";
```
[the remark this answers is not in the document]
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Files.Shares/src/ShareFileClient.cs:133]: Not sure, but I reverted this change.

File: sdk/storage/Azure.Storage.Files.Shares/tests/FileClientTests.cs [hunk 22]
```diff
@@ -6622,6 +6801,182 @@ public async Task RenameAsync_OAuth()
...
+            await using DisposingShare test = await SharesClientBuilder.GetTestShareAsync(nfs: true);
+            // Note that the parent directory was not created in this test case.
+            ShareDirectoryClient directory = InstrumentClient(test.Share.GetDirectoryClient(GetNewDirectoryName()));
+
+            ShareFileClient source = InstrumentClient(directory.GetFileClient(GetNewFileName()));
+            ShareFileClient hardLink = InstrumentClient(directory.GetFileClient(GetNewFileName()));
+
+            // Act
+            await TestHelper.AssertExpectedExceptionAsync<RequestFailedException>(
+                hardLink.CreateSymbolicLinkAsync(linkText: source.Uri.ToString()),
```
Comment (weirongw23-msft) [inline sdk/storage/Azure.Storage.Files.Shares/tests/FileClientTests.cs:6976]: Should this be `hardLink.CreateHardLinkAsync(linkText: source.Uri.ToString())` instead?

Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Files.Shares/tests/FileClientTests.cs]: Good catch!  Fixed!

**Review by weirongw23-msft — APPROVED** [review]

**Pull request merged**